README.md
# Znamky
[Czech] Jednoduchý správce známek z Bakalářů

Aplikace se přihlášuje jako uživatel do Bakalářů, čte jeho známky a další údaje a ukládá je na svůj server. Poté je možno je rychle a přehledně zobrazit a využít další šikovné nástroje.

Oficiální funkční aplikace je k nalezení na adrese: http://znamky.vsechno-atd.cz

Složky a speciální soubory:

/ obsahuje php soubory, které uživatel následně používá

/css/ obsahuje css styly

/img/ obsahuje veškerou grafiku aplikace

/js/ obsahuje javascript, který zajišťuje uživatelskou přívětivost aplikace

/php/ obsahuje php skripty volatelné javascriptem

/php/api.php budoucí jediný soubor pro komunikaci javascriptu se serverem

/php/c/ složka určená pro skripty cronu, které spouští server

/php/core/ složka se základními třídami aplikace (přenos jádra do baka.php není zatím dokončen!)

zatím jsou využívány i následující soubory, které by ale měly být časem přeneseny jinam:

/php/funkce.php základní používané php funkce

/php/projdiBakaweby.php hlavní cron, který aktuálně zajišťuje aktualizaci známek na pozadí

Aplikace zatím využívá jednotné přihlášení serveru Všechno-atd, které by mělo být brzy zrušeno.

Aplikace využívá jQuery a HTML5, je optimalizována pro prohlížeč Chrome a Firefox.

Data jsou ukládána do MySQL databáze, její struktura je přiložena.

Omluvte příšernou neuspořádanost kódu, ale začínal jsem to psát jako velice nezkušený programátor a teď se s tím těžko něco dělá, ale snad vidíte, že na tom pracuju :)
Nebojte se na cokoli zeptat na aslanex@vsechno-atd.cz


changelog.txt
5.2 (21-1-15)
-------------
*oprava nepřesného počítání průměru, pokud se nalezly nerozpoznatelné známky (ty se nyní nezapočítávají) a při desetinném průměru
*oprava zobrazení podrobností známek u anonymního zobrazení (zatím tu není k dispozici Když dostanu)

5.1 (13-1-15)
-------------
+vylepšení Když dostanu (lepší design, políčka nyní chtějí čísla, lepší čtení hodnoty známky)

5.0 (12-1-15)
-------------
+KDYŽ DOSTANU (odhadovač známky, dostupný po kliknuí na plus v detailu známek předmětu), bude se to ještě vylepšovat
-vypuštěno nepoužívané úsporné zobrazení známek, všem se nyní zobrazuje to normální (vertikální)

4.9 (15-11-14)
-------------
-zrušen nepoužívaný chat na hlavní stránce

4.8 (1-10-14)
-------------
-bakaláři se nepokouší zobrazit váhu, pokud není poprvé zobrazena (funkce dělala problémy)

4.7 (1-9-14)
-------------
+pokud nejsou při aktualizaci nalezeny žádné předměty se známkami, předměty účtu se promažou
*aktualizace enginu kvůli změně v bakalářích

4.6 (24-8-14)
-------------
oprava ukládání zdroje po anonymním zobrazení (zdroj se neukládal)

4.5 (19-6-14)
-------------
oprava kontroly nefunkčních zdrojů tak, aby se nefunkční přeskočily a kontrolovaly se další

4.4 (17-6-14)
-------------
drobné opravy manuální aktualizace, při neúspěšné aktualizaci se zobrazí chybové hlášení

4.3 (11-6-14)
-------------
trvalé přihlášení integrováno do Známek, nemělo by být potřeba klikat na přihlásit

4.2 (13-5-14)
-------------
nyní je každou půlhodinu kontrolováno, zda již nefunkční zdroje fungují (mělo by se tak předejít potřebné manuální aktualizaci při výpadku bakalářů)

4.1 (27-4-14)
-------------
počet pokusů o spojení s bakaláři zvýšen tak, aby to neotravovalo

4.0 (20-4-14)
-------------
na domovské stránce je nyní chat pro jednoduchou komunikaci s vývojářem (geniální nápad ©Aslanex)

3.8 (11-4-14)
-------------
úprava enginu, nyní se po neúspěšné synchronizaci zdroj dále nesynchronizuje, ale čeká na manuální aktualizaci (jeden nefunkční zdroj tak nezasekne všechny ostatní)

3.7 (22-3-14)
-------------
na hlavní stránce v tlačítku zdroje vedle klíče se nyní nachází tlačítko pro manuální aktualizaci známek, pokud jsou 3 hodiny a víc neaktualizované tak se to navíc červeně zobrazí

15-3-14
-------------
trvalé přihlášení funguje i na mobilech (nejde o aktualizaci Známek, ale jednotného přihlášení Všeho, atd)

3.6 (28-2-14)
-------------
[fix] aktualizace enginu kvůli dalšímu novému nesmyslnému zabezpečení v bakalářích
aktualizace FAQ

3.5 (25-2-14)
-------------
vylepšen parser adresy bakalářů

3.4 (8-2-14)
-------------
nové faviconky

3.3 (8-2-14)
-------------
pokud předmět nemá žádné známky, v podrobnosti předmětu se tato informace zobrazí (dočasné řešení, prázdné předměty by se neměly zobrazovat vůbec)

3.2 (7-2-14)
-------------
pokud je při synchronizaci u předmětu v databázi více známek než ve zdroji, přebytečné se nyní odstraní

3.1 (7-2-14)
-------------
server přestává zvládat automatickou synchronizaci - oprava enginu tak, aby nejprve synchronizoval naposledy nesynchronizované

3.0 (7-2-14)
-------------
přepracovaná domovská stránka pro přihlášené do stylu jako je ta pro nepřihlášené, přidání účtu nebo anonymní zobrazení je nyní také tak jednoduché

2.2 (3-2-14)
-------------
[fix] aktualizace enginu kvůli aktualizaci bakalářů v pololetí

2.1 (8-1-14)
-------------
[fix] zobrazení známek podporuje více roků

2.0 (23-12-13)
-------------
kompletně předělaný systém anonymního prohlížení, nová úvodní strana pro nepřihlášené a bez zdrojů, vše co nejjednodušší, nové logo
další drobné změny

1.7.1 (18-11-13)
-------------
[fix] zpět na výber školy ze seznamu při přidávání účtu

1.7 (16-11-13)
-------------
[feat] na stránce zobrazení známek se po najetí do pravého horního rohu údaje o uživateli změní v menu
[feat] kromě známek je nyní možno zjistit i výchovná opatření aktuálního pololetí, odkaz v menu při prohlížení známek
[fix] odkaz na znovupřidání účtu při neúspěšném pokusu funkční

1.6 (9-11-13)
-------------
[feat] jednoduchý tooltip návod pro nové uživatele

1.5 (8-11-13)
-------------
[feat] na hlavní straně se nyní zobrazují hlavní tlačítka zdrojů (dříve účtů) nad dvěma funkčními tlačítky
[feat] zobrazení známek by mělo být přehlednější (nyní strukturované)
[feat] výběr školy při přidávání zdroje (dříve účtu) je ještě jednodušší

1.4.5 (17-10-13)
-------------
[fix] odkaz zpět ve 3. fázi přidání účtu funguje

1.4.4 (15-10-13)
-------------
[fix] známky se skutečně zapisují i při uložení nového účtu

1.4.3 (30-9-13)
-------------
[fix] při průběžné kontrole známek se ukládá i nový průměr předmětu

1.4.2 (8-9-13)
-------------
[fix] průběžná kontrola známek funguje
[fix] známky se opět zapisují i při přidání účtu

1.4.1 (7-9-13)
-------------
[fix] průběžná kontrola známek, i pokud není přítomen checkbox na váhu

1.4 (6-9-13)
-------------
- při anonymním prohlížení je také oznámeno, pokud není žádná známka

1.3 (2-9-13)
-------------
- ve správě účtu je nyní možnost promazat známky, popsáno v FAQ

1.2 (1-9-13)
-------------
- pokud není žádná známka, je to oznámeno na výpisu známek

1.1 (27-8-13)
-------------
- nyní je možno si místo zadání adresy bakalářů vybrat ze seznamu škol
- podrobnější kontrola adresy bakalářů, pokud končí na "login.aspx"
- upraveny pokyny na zjištění adresy bakalářů v FAQ

1.0 (10-8-13)
-------------
- založen changelog
- na každé stránce je nyní vpravo dole svislý oddíl About, kde je odkaz na FAQ, changelog a číslo verze Známek


zobraz.php
<?
  require 'php/funkce.php';
  $funkce=new funkce();
  $funkce->mysqli->select_db('vsechno-atdcz_znamky');
  $bakaUcetID=(int) $_GET['bakaUcet'];
  if(!$_SESSION['ucetID'] || !$bakaUcetID) {
    header('HTTP/1.0 307 Retarded');
    header('Location: /');
    exit();
  }
  $funkce->zacatek('zobraz','Známky');
  if($funkce->mysqli->query('SELECT COUNT(*) FROM bakaUcty WHERE bakaUcetID="'.$bakaUcetID.'" AND ucetID='.$_SESSION['ucetID'])->fetch_assoc()['COUNT(*)']!=1) exit('Nejsi majitel tohoto účtu.');


  $ucet=$funkce->mysqli->query('SELECT jmeno,skola,zobrazeni FROM bakaUcty WHERE bakaUcetID="'.$bakaUcetID.'"')->fetch_assoc();
  $funkce->mysqli->query('UPDATE bakaUcty SET nove=0 WHERE bakaUcetID="'.$bakaUcetID.'"');
  echo '<div id="hlavicka" onmouseover="showZobrazMenu()" onmouseout="hideZobrazMenu()"><div id="zpet" onclick="if(event.which==1) location.href=\'/\'"></div><div id="skola">'.$ucet['skola'].'</div><div id="jmeno">'.$ucet['jmeno'].'</div><div id="zobrazMenu">skoč na<br><a href="javascript:void(0)" onclick="karta(1)" id="karta1link" style="color:#666">známky</a> - <a href="javascript:void(0)" onclick="karta(2)" id="karta2link">výchovná opatření</a> - <a href="spravuj?bakaUcet='.$bakaUcetID.'">správa</a></div></div><div id="obsahStranky"><div id="karta1" class="znamky karta">';

  $pololeti=(date('n')<1 || date('n')>7) ? 1 : 2;
  $zacatekPololeti=(date('n')==1) ? mktime(0,0,0,9,1,date('Y')-1) : ($pololeti==1) ? mktime(0,0,0,9,1) : mktime(0,0,0,2,1);
  $timelineSize=($pololeti==2) ? (date('n')-1)*150 : (date('n')==1) ? 750 : (date('n')-8)*150;

  $predmety=$funkce->mysqli->query('SELECT predmetID,nazev,prumer,ctvrtleti,pololeti FROM bakaPredmety WHERE bakaUcetID="'.$bakaUcetID.'" ORDER BY nazev');
  if($predmety->num_rows==0) echo 'Vypadá to, že nemáš žádné známky.<br>Uvědom si prosím, že aktualizace známek může trvat až 30 minut.';
  else {
    while($predmet=$predmety->fetch_assoc()) { // předměty

      $vysledna=($predmet['pololeti']) ? $predmet['pololeti'] : ( ($predmet['ctvrtleti']) ? $predmet['ctvrtleti'] : '&nbsp;' ) ;
      $vyslednaClass=($predmet['pololeti']) ? 'pololetniZnamka' : ( ($predmet['ctvrtleti']) ? 'ctvrtletniZnamka' : '' ) ;

      $znamky=$funkce->mysqli->query('SELECT znamka,vaha,datum,popis,nove FROM bakaZnamky WHERE predmetID="'.$predmet['predmetID'].'" ORDER BY datum');
      $funkce->mysqli->query('UPDATE bakaZnamky SET nove=NULL WHERE predmetID="'.$predmet['predmetID'].'"');
      /*if($ucet['zobrazeni']==1) {    // HORIZONTÁLNÍ ZOBRAZENÍ ZNÁMEK - ZATÍM VYPUŠTĚNO
        $poradiMinus=0;
        $lastXpozice=0; $Ypozice=7; $radek=1; $maxRadek=2;
        $znamkyA=null; $popisy=null; $vahy=null; $fdatumy=null; $delky=null; $Xpozice=null; $radky=null;
        for($c=0;$znamka=$znamky->fetch_assoc();$c++) {
          $fontSize=(!$znamka['vaha']) ? 6 : $znamka['vaha'];
          $nova=($znamka['nove']==1) ? ';color:firebrick' : '';
          echo '<article style="font-size:'.(20+$fontSize*2).'px'.$nova.'">'.$znamka['znamka'].'</article>';

          $poradi=$c-$poradiMinus;
          if($lastPopis==$znamka['popis'] && $lastDatum==$znamka['datum'] && $lastVaha==$znamka['vaha']) {
            $poradi--; $poradiMinus++;
            $znamkyA[$poradi].=', '.$znamka['znamka'];
            $delky[$poradi]+=10;
            }
          else {
            $znamkyA[$poradi]=$znamka['znamka'];
            $popisy[$poradi]=$znamka['popis'];
            $vahy[$poradi]=$znamka['vaha'];
            $fdatumy[$poradi]=date('j.n.y',$znamka['datum']);
            $nove[$poradi]=$znamka['nove'];
            $delky[$poradi]=(strlen($popisy[$poradi])*8)+90;
            $lastPopis=$znamka['popis']; $lastDatum=$znamka['datum']; $lastVaha=$znamka['vaha'];
            $Xpozice[$poradi]=($znamka['datum']-$zacatekPololeti)/17280;
            $radky[$poradi]=radekZnamky($delky,$radky,$Xpozice,$Xpozice[$poradi]);
            if($radky[$poradi]>$maxRadek) {
              $maxRadek=$radky[$poradi];
              if((strlen($predmet['nazev'])*12)-130>$Xpozice[$poradi]) $maxRadek++;
              }
            $lastXpozice=$Xpozice[$poradi];
            }
          }

        $ctvrtletniPodrobnost=($predmet['pololeti'] && $predmet['ctvrtleti']) ? '<div class="ctvrtletniPodrobnost" style="bottom:'.(($maxRadek*16)-18).'px">čtvrtletí: '.$predmet['ctvrtleti'].'</div>' : '';
        echo '</div><div class="podrobnostiPredmetu1" id="podrobnostiPredmetu'.$predmet['predmetID'].'" style="margin-top: '.(($maxRadek*16)+8).'px; display: none; opacity: 0;">'.$ctvrtletniPodrobnost.'<img src="/img/blank" class="timeline" style="background-image: url(\'/img/cara'.$pololeti.'\');width: '.$timelineSize.'px">';

        foreach($znamkyA as $c=>$a) {
          $nova=($nove[$c]==1) ? ' style="color:firebrick"' : '';
          $vaha=($vahy[$c]) ? ', váha '.$vahy[$c] : null;
          echo '<article style="margin-left:'.$Xpozice[$c].'px;top:-'.(($radky[$c]*16)-9).'px"><b'.$nova.'>'.$a.'</b> '.$popisy[$c].' <span>'.$fdatumy[$c].$vaha.'</span></article>';
          }
        echo '</div><script>predmetHeight['.$predmet['predmetID'].']='.(($maxRadek*16)+38).'</script></section>';
        }*/
      // vertikální zobrazení známek (aktuálně jediné)
      $znamkyA=array();
      $poradiMinus=0;
      $vypisHtml='';
      $znamkyTotal=0; $vahyTotal=0; $znameVahu=false; $znameZnamky=true;
      for($c=0;$znamka=$znamky->fetch_assoc();$c++) {
        $fontSize=(!$znamka['vaha']) ? 6 : $znamka['vaha'];
        $nova=($znamka['nove']==1) ? ';color:firebrick' : '';
        $vypisHtml.='<article style="font-size:'.(20+$fontSize*2).'px'.$nova.'">'.$znamka['znamka'].'</article>';

        switch($znamka['znamka']) {
          case '1-': $hodnotaZnamky = 1.5; break;
          case '2-': $hodnotaZnamky = 2.5; break;
          case '3-': $hodnotaZnamky = 3.5; break;
          case '4-': $hodnotaZnamky = 4.5; break;
          case '1': case '2': case '3': case '4': case '5':
            $hodnotaZnamky = (int) $znamka['znamka']; break;
          default: $hodnotaZnamky = false;
        }
        if($hodnotaZnamky!=false) {
          $znamkyTotal += ($znamka['vaha']) ? $hodnotaZnamky*$znamka['vaha'] : $hodnotaZnamky*10;
          if($znamka['vaha']) {
            $vahyTotal += $znamka['vaha'];
            $znameVahu = true;
          }
          else
            $vahyTotal += 10;
        }
        else
          $znameZnamky = false;

        $poradi=$c-$poradiMinus;
        if($lastPopis==$znamka['popis'] && $lastDatum==$znamka['datum'] && $lastVaha==$znamka['vaha']) {
          $poradi--; $poradiMinus++;
          $znamkyA[$poradi]->znamka.=', '.$znamka['znamka'];
        }
        else {
          $znamkyA[$poradi]=new StdClass;
          $znamkyA[$poradi]->znamka=$znamka['znamka'];
          $znamkyA[$poradi]->popis=$znamka['popis'];
          $znamkyA[$poradi]->vaha=$znamka['vaha'];
          $znamkyA[$poradi]->rok=($znamka['datum']) ? date('Y',$znamka['datum']) : -1;
          $znamkyA[$poradi]->mesic=($znamka['datum']) ? date('n',$znamka['datum']) : -1;
          $znamkyA[$poradi]->fdatum=($znamka['datum']) ? date('j.n.y',$znamka['datum']) : 'bez data';
          $znamkyA[$poradi]->nove=$znamka['nove'];
          $lastPopis=$znamkyA[$poradi]->popis; $lastVaha=$znamkyA[$poradi]->vaha; $lastDatum=$znamka['datum'];
        }
      }

      $ctvrtletniPodrobnost=($predmet['pololeti'] && $predmet['ctvrtleti']) ? '<div class="ctvrtletniPodrobnost">čtvrtletí: '.$predmet['ctvrtleti'].'</div>' : '';
      $podrobnostiHtml = '</div><div class="podrobnostiPredmetu2" id="podrobnostiPredmetu'.$predmet['predmetID'].'">'.$ctvrtletniPodrobnost.'<div class="podrobnostiPredmetu2Znamky" onclick="event.stopPropagation()">';

      if(count($znamkyA)===0) {
        $detailHtml = '<article style="color:#444;opacity:0.5">žádné známky</article>';
        $articleHeight = 61;
      }
      else {
        $mesic=-2; $mesicC=0; $rok=$znamkyA[0]->rok; $detailHtml='';
        foreach($znamkyA as $c=>$a) {
          if($a->rok>$rok) {
            $mesic=-1;
            $rok=$a->rok;
          }
          if($a->mesic>$mesic) {
            $detailHtml.='<div class="znamkyMesic">'.$funkce->month[($a->mesic-1)].'</div>';
            $mesic=$a->mesic; $mesicC++;
            $margin=' style="margin-top:4px"';
          }
          else
            $margin='';
          $nova=($a->nove==1) ? ' style="color:firebrick"' : '';
          $vaha=($a->vaha) ? ', v'.$a->vaha : '';
          $detailHtml.='<article'.$margin.'><b'.$nova.'>'.$a->znamka.'</b> '.$a->popis.' <span>'.$a->fdatum.$vaha.'</span></article>';
        }
        $detailHtml.='</div><div class="pridejZnamku"><div class="znamkyMesic" id="pridejZnamkuMesic'.$predmet['predmetID'].'" style="display:none;opacity:0" onclick="event.stopPropagation();document.querySelector(\'#pridejZnamkuInput'.$predmet['predmetID'].'\').select()">když dostanu</div><article style="margin-top:4px"><a href="javascript:void(1);" onclick="event.stopPropagation();pridejZnamku('.$predmet['predmetID'].')" class="pridejZnamkuA" id="pridejZnamkuA'.$predmet['predmetID'].'"></a><span style="display:none;" id="pridejZnamkuCont'.$predmet['predmetID'].'" onclick="event.stopPropagation();document.querySelector(\'#pridejZnamkuInput'.$predmet['predmetID'].'\').select()"><input type="tel" onchange="vypocitejNovouZnamku('.$predmet['predmetID'].')" onkeyup="vypocitejNovouZnamku('.$predmet['predmetID'].')" onclick="event.stopPropagation();this.select();" id="pridejZnamkuInput'.$predmet['predmetID'].'" class="pridejZnamkuInput" value="5" maxlength="2"> váhy <input type="number" min="1" max="20" onchange="vypocitejNovouZnamku('.$predmet['predmetID'].')" onkeyup="vypocitejNovouZnamku('.$predmet['predmetID'].')" onclick="event.stopPropagation();this.select();" id="pridejZnamkuVaha'.$predmet['predmetID'].'" class="pridejZnamkuInput" value="10">, průměr bude <span class="vypocitanyPrumer" id="pridejZnamkuVysledek'.$predmet['predmetID'].'"></span> </span></div></article>';
        $articleHeight = (count($znamkyA)*17+$mesicC*4+65);
      }
      $dataVahy = ($znameVahu==false) ? 0 : 1;
      $dataZnamky = ($znameZnamky==false) ? 0 : 1;

      echo '<section onclick="podrobnostiPredmetu('.$predmet['predmetID'].')" id="sectionPredmetu'.$predmet['predmetID'].'" data-grades-total='.$znamkyTotal.' data-grades-ammount='.$vahyTotal.' data-grades-isznamky='.$dataZnamky.' data-grades-isvahy='.$dataVahy.' data-article-height='.$articleHeight.'><div class="nazevPredmetu">'.$predmet['nazev'].'</div><div class="vyslednaPredmetu '.$vyslednaClass.'">'.$vysledna.'</div><div class="prumerPredmetu">'.$predmet['prumer'].'</div><div class="znamkyPredmetu" id="znamkyPredmetu'.$predmet['predmetID'].'">'.$vypisHtml.$podrobnostiHtml.$detailHtml.'</div></section>';
    }

  }

    echo '</div><div id="karta2" class="vychOpatreni karta">';

    $vychOpatreni=$funkce->mysqli->query('SELECT opatreniID,druh,datum,text FROM bakaVychOpatreni WHERE bakaUcetID="'.$bakaUcetID.'" ORDER BY datum');  // výchovná opaření
    if($vychOpatreni->num_rows==0) echo 'Vypadá to, že nemáš žádná výchovná opatření.<br>Uvědom si prosím, že aktualizace může trvat až 30 minut.';
    else {
      while($opatreni=$vychOpatreni->fetch_assoc()) {

        $datum=date('j. ',$opatreni['datum']).$funkce->month[(int) date('n',$opatreni['datum'])];
        echo '<section><div class="druh">'.$opatreni['druh'].'</div><div class="datum">'.$datum.'</div><div class="text">'.$opatreni['text'].'</div></section>';

      }
    }
    echo '</div>';

  $funkce->htmlKonec();

  function radekZnamky($delky,$radky,$Xpozice,$Xpoz,$radek=1) {
    foreach($delky as $c=>$a) {
      if($radky[$c]==$radek && $a+$Xpozice[$c]>$Xpoz) {
        $radek++;
        $radek=radekZnamky($delky,$radky,$Xpozice,$Xpoz,$radek);
      }
    }
    return $radek;
  }
?>
